src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5532,6 +5532,107 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
         cbnzw(tmp5, DONE);
       }
     }
+  } else if (UseCompactObjectHeaders) {
+    Label NEXT_64BYTE, NEXT_16BYTE, TAIL, TAIL03, TAIL01;
+    mov(result, false);
+    // if (a1 == null || a2 == null)
+    //     return false;
+    cbz(a1, DONE);
+    ldrw(cnt1, Address(a1, length_offset));
+    cbz(a2, DONE);
+    // if (a1.length != a2.length)
+    //      return false;
+    ldrw(cnt2, Address(a2, length_offset));
+    cmpw(cnt2, cnt1);
+    br(NE, DONE);
+    // if (a1.length == 0)
+    //      return true;
+    cbz(cnt1, SAME);
+    lea(a1, Address(a1, base_offset));
+    lea(a2, Address(a2, base_offset));
+
+    cmpw(cnt1, elem_per_word);
+    if (UseSIMDForArrayEquals) {
```

Comment:
highly-probable case ([8, 64) bytes) needs two jumps, need reordering

the default implementation only requires two fall-throughs

Reply:
done